### ISSUE TITLE
A way to make breaking changes harmless!

### ISSUE THREAD

**Issue by MohammadHamdyGhanem** [dotnet/roslyn#29022]
Sometimes, we ask for new features or correcting some behavior, but we get the "It will be a breaking change" response!
I agree that breaking changes are frustrating. So, I think of a compromise:
Add a compiler option for each breaking feature so we can turn it off at the project level (with a tab to do that in the project properties window) and/or at the any code page level. 
This way, if we have an old project and want it to run on a new .net version that has breaking changes, we can choose to turn them off so the project can run as expected. Therefore, we can write new projects using the new features, without breaking any existing projects.
This is about changes in compiler behavior, but breaking changes in class libraries can be done safely using NuGets versions. If there is a breaking change in some API, you can provide the old API as a NuGet.
In short: I don't see any reason to kill good ideas or not to fix past mistakes just because fearing of breaking changes. There is always a solution, and I think this is it.

**Comment by CyrusNajmabadi** [comment]
This is now very painful for the entire ecosystem.  Now, instead of just saying "here's some code, you need this version of the language to run it" you now have to also know the exact set of breaking changes options that need to be specified for the code to run.

On top of that, you've now greatly increased the complexity of the compiler (and IDE and so on).  For each of those options, you now need to test every single other feature.  You also now need to test each and every permutation of those features.  i.e. if user has accepted breaking change 1, 3, 5, and 7, but not 2, 4, 6, or 8.  You need to try those all out together and make sure they still work.

> There is always a solution, and I think this is it.

A very effective solutoin is also what Roslyn currently does which is to:

1. not make breaking changes by default.
2. when breaking changes are made, do so extremely carefully and conscientiously.

**Comment by MohammadHamdyGhanem** [comment]
@CyrusNajmabadi 
Don't forget that new developers can find dot net inhomogeneous because they have to write code in different styles because of earlier versions computability. You walked the pass from its beginning so you are familiar with this, but any new learner will find himself lost every time he has unexpected errors or behaviors (non-generic collections cause some of them as I showed in other issues).
A well designed programming language syntax should work as expected in all situations. It is not a programmers responsibility to memorize all exceptions and workaround just because some older programmer have a project that must work on the new version!
I think you should concentrate on attracting new developers because they are the future, and gave old fox temp solutions to stay in business.

**Comment by CyrusNajmabadi** [comment]
> You will know them just by compiling the project and seeing the errors. IDE can offer a fix for these errors that shows the right compiler option.

If it's a breaking change, you wouldn't have errors.  i.e. breaking changes occur when previously allowed *changes meaning*.  So the code compiles with the option set, or the option unset, except it now has different meaning.  In order for this program to run successfully, the person will have to tell you the set of breaks they enabled/disabled.

**Comment by CyrusNajmabadi** [comment]
@MohammadHamdyGhanem you also did not address this point:

> On top of that, you've now greatly increased the complexity of the compiler (and IDE and so on). For each of those options, you now need to test every single other feature. You also now need to test each and every permutation of those features. i.e. if user has accepted breaking change 1, 3, 5, and 7, but not 2, 4, 6, or 8. You need to try those all out together and make sure they still work.

**Comment by CyrusNajmabadi** [comment]
> but any new learner will find himself lost every time he has unexpected errors or behaviors (non-generic collections cause some of them as I showed in other issues).

It is far easier and cheaper to teach a 'new learner' how the language works than it is to break code that users care about, or maintain a compiler that attempts to support options for all these breaks.

--

If you don't believe me, i would recommend you actually try this out.  Attempt to make a few breaking changes in Roslyn and then validate all those options.  Remember that each of those options likely needs to be validated against each other as well.  

Try this out and see how expensive this is.  Until you work through all the costs and complexities involved here, it's not acceptable to just say you think this is a solution.  Solutions aren't solutions if they're hugely expensive, or would actually cause more problems and more difficulties than the problems they purport to solve.

**Comment by MohammadHamdyGhanem** [comment]
@CyrusNajmabadi 
I faced the breaking changes when I  moved from VB6 to VB.NET. this was a total dsstruction not just a break! But I gained a lot from VB.NET. If you keep refusing to do small breaks, you will someday make a huge movement like what happened from VB6 to VB.NET.
By the way: I have a VC++ project that works only on .NET 2013 but fails on higher versions becuase some breaking changes! 
It seems that MS does breaking changes all the time!

**Comment by svick** [comment]
> If you keep refusing to do small breaks, you will someday make a huge movement like what happened from VB6 to VB.NET.

Is that a bad thing? It seemed to work reasonably well for VB.

> I have a VC++ project that works only on .NET 2013 but fails on higher versions becuase some breaking changes!

VC++ is in a much more complicated position than Roslyn. They have an independent standard to implement (ISO C++), had issues with following that standard in the past and are getting better. That itself will require some breaking changes. On top of that, the standard itself also has its own breaking changes. So I think it's expected that VC++ would have more breaking changes than Roslyn.

> It seems that MS does breaking changes all the time!

They do. There's even [a fairly long list of breaking changes since C# 7.0](https://github.com/dotnet/roslyn/blob/master/docs/compilers/CSharp/Compiler%20Breaking%20Changes%20-%20post%20VS2017.md). But they are very careful about making them and about making sure they are worth it. And most of them are actually just fixing bugs, not reversing intentional design decisions.

**Comment by sharwell** [comment]
> Add a compiler option for each breaking feature so we can turn it off at the project level (with a tab to do that in the project properties window) and/or at the any code page level.

Feature flags sound like a good idea, but they are not sustainable. Each flag doubles the number of language variants that need to be tested, so the overall impact of each feature on the test suite goes from O(1) to O(2<sup>n</sup>). In practice, this quickly breaks down as users start to find countless cases where their unique set of options isn't working the way they expected. It also makes it impossible for the language standardization committee to fully evaluate and define the current language.

I'm going to close this as **External** since discussions of the progression of the C# language belong on [dotnet/csharplang](https://github.com/dotnet/csharplang), but it would be reasonable to equate this with **Won't Fix** there is no known viable maintenance path forward.
